engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -880,6 +883,20 @@ private void execute() throws YarnException, IOException
         if (0 != exitStatus) {
           if (allocatedContainer != null) {
             numFailedContainers.incrementAndGet();
+            if (exitStatus != 1) {
+              // If container failure due to framework
+              AtomicInteger failedTimes = new AtomicInteger(1);
+              String hostname = allocatedContainer.container.getNodeId().getHost();
+              if (failedContainersMap.containsKey(hostname)) {
```

Comment:
Using ConcurrentMap does not prevent thread switch here. In case failedContainersMap does not have mapping for hostname, there is race condition here. You may use putIfAbsent() if the map can be accessed from two or more different threads.